pkg/engine/project.go

```diff
@@ -50,16 +49,20 @@ func getPwdMain(root, main string) (string, string, error) {
 	if main == "" {
 		main = "."
 	} else {
-		// The path must be relative from the package root.
-		if path.IsAbs(main) {
-			return "", "", errors.New("project 'main' must be a relative path")
-		}
+		// Note: I had to relax constraints on main being a subfolder of root. I tried computing a relative
+		// path from the auto api tmp directory. But the tmp directory on osx is `/private/var/tmp` and
+		// computing a relative path doesn't work as `/private/User/...` (where the program would be) isn't
+		// a valid file path.
+		//
+		// The reason we had this check in the first place is due to PPC constraints for hosted execution. Look through the original PR
+		// https://github.com/pulumi/pulumi/pull/615 https://github.com/pulumi/pulumi-ppc/pull/102
+		// It would appear that it is now safe to allow root and main to be in disjoint file trees.
+		// Empyrically from my testing it seems to work, but would be worth checking with @chrsmith
```

Comment:
FWIW - this matches my understanding as well.  The original motivation for the constraint was that we needed to be able to zip up the folder containing the `Pulumi.yaml` file and have that zip define the full source context for the deployment.  This is no longer a formal scenario.

That said - it is certainly possible that we developed other implicit dependencies on this requirement elsewhere because this has been a guarantee.  I cannot think of any likely places to be concerned about though (closure serialization came to mind, but it keys off of `package.json` location, not `Pulumi.yaml` location).

Reply:
Yes, we had the "need to zip everything we'll need up for the PPC" scenario. So any PPC-specific requirements can be ignored.

I don't recall any other invariants we needed to maintain in this area. The only thing I do recall, is support for adding the `main` field to `Pulumi.yaml` to support "up and over" references. (See https://github.com/pulumi/pulumi/pull/582.) ... but that seems like it is essentially what you are trying to enable here.